pkg/agent/plugin/workloadattestor/docker/docker.go

```diff
@@ -0,0 +1,161 @@
+package docker
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"sync"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	dockerclient "github.com/docker/docker/client"
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire/pkg/agent/common/cgroups"
+	"github.com/spiffe/spire/proto/agent/workloadattestor"
+	"github.com/spiffe/spire/proto/common"
+	spi "github.com/spiffe/spire/proto/common/plugin"
+)
+
+const (
+	selectorType        = "docker"
+	subselectorLabel    = "label"
+	subselectorImageID  = "image_id"
+	defaultCgroupPrefix = "/docker"
+)
+
+var defaultContainerIndex = 1
+
+// DockerClient is a subset of the docker client functionality, useful for mocking.
+type DockerClient interface {
+	ContainerInspect(ctx context.Context, containerID string) (types.ContainerJSON, error)
+}
+
+type dockerPlugin struct {
+	docker               DockerClient
+	cgroupPrefix         string
+	cgroupContainerIndex int
+	fs                   cgroups.FileSystem
+	mtx                  *sync.RWMutex
+}
+
+type dockerPluginConfig struct {
+	// DockerSocketPath is the location of the docker daemon socket (default: "unix:///var/run/docker.sock" on unix).
+	DockerSocketPath string `hcl:"docker_socket_path"`
+	// DockerVersion is the API version of the docker daemon (default: "1.40").
+	DockerVersion string `hcl:"docker_version"`
+	// CgroupPrefix is the cgroup prefix to look for in the cgroup entries (default: "/docker").
+	CgroupPrefix string `hcl:"cgroup_prefix"`
+	// CgroupContainerIndex is the index within the cgroup path where the container ID should be found (default: 1).
+	// This is a *int to allow differentiation between the default int value (0) and the absence of the field.
+	CgroupContainerIndex *int `hcl:"cgroup_container_index"`
+}
+
+func (p *dockerPlugin) Attest(ctx context.Context, req *workloadattestor.AttestRequest) (*workloadattestor.AttestResponse, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	cgroupList, err := cgroups.GetCgroups(req.Pid, p.fs)
+	if err != nil {
+		return nil, err
+	}
+
+	var containerID string
+	var hasDockerEntries bool
+	for _, cgroup := range cgroupList {
+		// We are only interested in cgroup entries that match our desired prefix. Example entry:
+		// "10:perf_event:/docker/2235ebefd9babe0dde4df4e7c49708e24fb31fb851edea55c0ee29a18273cdf4"
+		if !strings.HasPrefix(cgroup.GroupPath, p.cgroupPrefix) {
+			continue
+		}
+		hasDockerEntries = true
+
+		parts := strings.Split(cgroup.GroupPath, "/")
+
+		if len(parts) <= p.cgroupContainerIndex+1 {
+			log.Printf("Docker entry found, but is missing the container id: %v", cgroup.GroupPath)
+			continue
+		}
+		containerID = parts[p.cgroupContainerIndex+1]
+		break
+	}
+
+	// Not a docker workload. Since it is expected that non-docker workloads will call the
+	// workload API, it is fine to return a response without any selectors.
```

Comment:
@APTy I'm very interested in exploring how to use spire with docker, so thanks for this plugin. I'm new to spire, and when I read the above, I'm a little bit confused. I'd appreciate if you can offer some help.

Does this `workloadattestor` run (i.e., the command `spire-agent run`) inside the container or reside in the docker host?